stonesoup/dataassociator/neighbour.py

```diff
@@ -106,3 +110,108 @@ def associate(self, tracks, detections, time):
         associations = max(joint_hypotheses)
 
         return associations
+
+
+class GNNWith2DAssignment(DataAssociator):
+    """Global Nearest Neighbour Associator
+
+    Associates detections to a predicted state using the
+    Global Nearest Neighbour method, utilising a 2D matrix of
+    distances and a "shortest path" assignment algorithm.
+    """
+
+    hypothesiser = Property(
+        Hypothesiser,
+        doc="Generate a set of hypotheses for each prediction-detection pair")
+
+    def associate(self, tracks, detections, time):
+        """Associate a set of detections with predicted states.
+
+        Parameters
+        ----------
+        tracks : set of :class:`Track`
+            Current tracked objects
+        detections : set of :class:`Detection`
+            Retrieved measurements
+        time : datetime
+            Detection time to predict to
+
+        Returns
+        -------
+        dict
+            Key value pair of tracks with associated detection
+        """
+
+        associations = {}
+
+        # Convert sets to indexable lists
+        tracks = list(tracks)
+        detections = list(detections)
+
+        # Generate 2d array "matrix" of hypotheses mapping track to detection
+        hypothesis_matrix = np.empty(
```

Comment:
Further speed improvements can be achieved by only performing data association for tracks that are detected (i.e. not missed), as the 2D assignment complexity scales according the number of targets (evident from the shape requirement of `hypothesis_matrix` and `distance_matrix`).

This could be achieved by replacing `tracks = list(tracks)` with:
```python
# Extract detected tracks
# NOTE: It is assumed that if a track is undetected/missed, then it will only have a single hypothesis, which
#       is also a miss-detection hypothesis (i.e. evaluates to False)
detected_tracks = [track for track in tracks if not(len(hypotheses[track]) == 1 and not hypotheses[track][0])]

# Store associations for undetected/missed tracks
for track in tracks-set(detected_tracks):
    associations[track] = hypotheses[track][0]

# No need to perform data association if all tracks are missed
if not len(detected_tracks):
    return associations
```
and then replacing all occurrences of `tracks` with `detected tracks` further down.

The above assumes that hypothesis generation has been performed as suggested in my other comment.